src/Workspaces/Remote/ServiceHub/Services/CustomMessageHandler/RemoteCustomMessageHandlerService.cs

```diff
@@ -0,0 +1,252 @@
+            => new RemoteCustomMessageHandlerService(arguments);
+    }
+
+    public RemoteCustomMessageHandlerService(in ServiceConstructionArguments arguments)
+        : base(arguments)
+    {
+    }
+
+    public ValueTask<HandleCustomMessageResponse> HandleCustomMessageAsync(
+        Checksum solutionChecksum,
+        string assemblyPath,
+        string typeFullName,
+        string jsonMessage,
+        DocumentId? documentId,
+        ImmutableArray<LinePosition> positions,
+        CancellationToken cancellationToken)
+    {
+#if NETSTANDARD2_0
+        throw new InvalidOperationException("Custom handlers are only supported in the servicehub host");
+#else
+
+#if DEBUG
+        System.Diagnostics.Debugger.Launch();
+#endif
+
+        Requires.NotNullOrEmpty(assemblyPath);
+        Requires.NotNullOrEmpty(typeFullName);
+        Requires.NotNullOrEmpty(jsonMessage);
+        Requires.NotNullOrEmpty(positions);
+
+        return RunServiceAsync(solutionChecksum, async solution =>
+        {
+            Document? document = null;
+            if (documentId is not null)
+            {
+                document = solution.GetDocument(documentId) ?? await solution.GetSourceGeneratedDocumentAsync(documentId, cancellationToken).ConfigureAwait(false);
+                Contract.ThrowIfNull(document);
+            }
+
+            try
+            {
+                var extension = _extensions.GetOrAdd(assemblyPath, _ =>
+                {
+                    var directory = Path.GetDirectoryName(assemblyPath);
+                    Contract.ThrowIfNull(directory);
+
+                    var loadContext = new System.Runtime.Loader.AssemblyLoadContext(name: $"RemoteCustomMessageHandlerService assembly load context for {assemblyPath}", isCollectible: true);
+                    loadContext.Resolving += (context, assemblyName) =>
+                    {
+                        var sharedAssembly = _defaultLoadContext.Assemblies.Where(a => a.GetName().Name == assemblyName.Name).FirstOrDefault();
+
+                        if (sharedAssembly is not null)
+                        {
+                            if (sharedAssembly.GetName().Version < assemblyName.Version)
+                            {
+                                throw new InvalidOperationException($"The version of the loaded assembly {assemblyName.Name} is too low: requested {assemblyName.Version}, found {sharedAssembly.GetName().Version}.");
+                            }
+
+                            return sharedAssembly;
+                        }
+
+                        var extensionAssemblyPath = Path.Combine(directory, $"{assemblyName.Name}.dll");
+                        if (File.Exists(extensionAssemblyPath))
+                        {
+                            return loadContext.LoadFromAssemblyPath(extensionAssemblyPath);
+                        }
+
+                        return null;
```

Comment:
I think this should not be returning null, and instead throwing a file not found.  But I believe you could just call `LoadFromAssemblyPath` without checking the path which will throw file not found.